### PR TITLE
Fix kind of (->) and introduce FUN

### PR DESCRIPTION
**Pull request by monoidal**
Refs #204

### CODE REVIEW

File: compiler/hsSyn/Convert.hs [hunk 1]
```diff
@@ -1334,7 +1334,7 @@ cvtTypeKind ty_str ty
                           _                -> return x'
                  returnL (HsFunTy noExt x'' HsOmega y')
              | otherwise ->
-                  mk_apps (HsTyVar noExt NotPromoted (noLoc (getRdrName funTyCon))) -- TODO: arnaud: fix when the above is done
+                  mk_apps (HsTyVar noExt NotPromoted (noLoc (getRdrName arrowTyCon))) -- TODO: arnaud: fix when the above is done
```
Comment (aspiwack) [inline compiler/hsSyn/Convert.hs]:
```suggestion
                  mk_apps (HsTyVar noExt NotPromoted (noLoc (getRdrName arrowTyCon))
```

File: compiler/hsSyn/HsTypes.hs [hunk 2]
```diff
@@ -1097,6 +1097,8 @@ mkHsAppTys = foldl' mkHsAppTy
 --      splitHsFunType (a -> (b -> c)) = ([a,b], c)
 -- Also deals with (->) t1 t2; that is why it only works on LHsType Name
 --   (see Trac #9096)
+--
+-- TODO Krzysztof: we might allow FUN for consistency.
```
Comment (aspiwack) [inline compiler/hsSyn/HsTypes.hs]: TODO needs to disappear from the final merge. Turn it into an issue before submitting.

File: compiler/iface/IfaceType.hs [hunk 3]
```diff
@@ -1203,6 +1203,11 @@ pprTyTcApp' ctxt_prec tc tys dflags style
   , rep `ifaceTyConHasKey` liftedRepDataConKey
   = kindType
 
+  | tc `ifaceTyConHasKey` funTyConKey
+  , IA_Vis (IfaceTyConApp rep IA_Nil) args <- tys
+  , rep `ifaceTyConHasKey` omegaDataConKey
+  = pprIfacePrefixApp ctxt_prec (parens arrow) (map (ppr_ty appPrec) (tys_wo_kinds_fun args))
```
Comment (aspiwack) [inline compiler/iface/IfaceType.hs]: Do you know why this special casing needs to be?

File: compiler/prelude/primops.txt.pp [hunk 4]
```diff
@@ -161,15 +161,15 @@
 -- This type won't be exported directly (since there is no concrete
 -- syntax for this sort of export) so we'll have to manually patch
 -- export lists in both GHC and Haddock.
-primtype (->) a b
-  {The builtin function type, written in infix form as {\tt a -> b} and
+primtype FUN a b
+  {TODO update
```
Comment (aspiwack) [inline compiler/prelude/primops.txt.pp]: Fix this TODO before merging.

File: compiler/prelude/primops.txt.pp [hunk 5]
```diff
@@ -161,15 +161,15 @@
...
+   The builtin function type, written in infix form as {\tt a -> b} and
    in prefix form as {\tt (->) a b}. Values of this type are functions
    taking inputs of type {\tt a} and producing outputs of type {\tt b}.
 
    Note that {\tt a -> b} permits levity-polymorphism in both {\tt a} and
    {\tt b}, so that types like {\tt Int\# -> Int\#} can still be well-kinded.
   }
-  with fixity = infixr -1
          -- This fixity is only the one picked up by Haddock. If you
          -- change this, do update 'ghcPrimIface' in 'LoadIface.hs'.
```
Comment (aspiwack) [inline compiler/prelude/primops.txt.pp]: Remove the comment if you are removing the fixity thing.

File: compiler/typecheck/TcValidity.hs [hunk 6]
```diff
@@ -1273,17 +1273,24 @@ tcInstHeadTyNotSynonym :: Type -> Bool
...
+-- But we allow
+-- 1) kind instantiations
+-- 2) the type (->) = FUN 'Omega, even though it's not in this form.
 tcInstHeadTyAppAllTyVars ty
   | Just (tc, tys) <- tcSplitTyConApp_maybe (dropCasts ty)
-  = ok (filterOutInvisibleTypes tc tys)  -- avoid kinds
+  = let tys' = filterOutInvisibleTypes tc tys  -- avoid kinds
+        tys'' | tc == funTyCon, tys_h:tys_t <- tys', tys_h `eqType` omegaDataConTy = tys_t
+              | otherwise = tys'
+    in ok tys''
```
Comment (aspiwack) [inline compiler/typecheck/TcValidity.hs]: Do we need to special case here as well as in `tcInstHeadTyNotSynonym`, though? If so :heavy_check_mark:

File: testsuite/tests/th/T14888.stderr [hunk 7]
```diff
@@ -7,4 +7,4 @@ T14888.hs:18:23-59: Splicing expression
     "class T14888.Functor' (f_0 :: * -> *)
     where T14888.fmap' :: forall (a_1 :: *) (b_2 :: *) .
                           (a_1 -> b_2) -> f_0 a_1 -> f_0 b_2
-instance T14888.Functor' ((->) r_3 :: * -> *)"
+instance T14888.Functor' ((->) r_3)"
```
Comment (aspiwack) [inline testsuite/tests/th/T14888.stderr]: This is kind of weird :slightly_smiling_face:

File: testsuite/tests/typecheck/should_compile/abstract_refinement_hole_fits.stderr [hunk 8]
```diff
@@ -240,14 +240,14 @@ abstract_refinement_hole_fits.hs:7:5: warning: [-Wtyped-holes (in -Wdefault)]
...
           where fail :: forall (m :: * -> *) a. Monad m => String -> m a
         return (_ :: Integer -> [Integer] -> Integer) (_ :: t1)
           where return :: forall (m :: * -> *) a. Monad m => a -> m a
         pure (_ :: Integer -> [Integer] -> Integer) (_ :: t1)
           where pure :: forall (f :: * -> *) a. Applicative f => a -> f a
-        uncurry (_ :: a4 -> b3 -> Integer -> [Integer] -> Integer)
-                (_ :: (a4, b3))
-          where uncurry :: forall a b c. (a -> b -> c) -> (a, b) -> c
-        ($!) (_ :: a6 -> Integer -> [Integer] -> Integer) (_ :: a6)
-          where ($!) :: forall a b. (a -> b) -> a -> b
```
Comment (aspiwack) [inline testsuite/tests/typecheck/should_compile/abstract_refinement_hole_fits.stderr]: This whole file (and the following few) looks unrelated (also weird!)

File: compiler/typecheck/TcValidity.hs (hunk 6 above)
Comment (monoidal) [reply compiler/typecheck/TcValidity.hs]: Both are needed, one is for `FlexibleInstances`, one for `TypeSynonymInstances`.

File: testsuite/tests/typecheck/should_compile/abstract_refinement_hole_fits.stderr (hunk 8 above)
Comment (monoidal) [reply testsuite/tests/typecheck/should_compile/abstract_refinement_hole_fits.stderr]: The ordering in hole tests seems to vary depending on internals, and it seems that it changed together with `(->)`. Ideally, it should be sorted, but I don't want to change this in this PR.

Comment (aspiwack) [reply testsuite/tests/typecheck/should_compile/abstract_refinement_hole_fits.stderr]: Absolutely.

File: compiler/iface/IfaceType.hs (hunk 3 above)
Comment (monoidal) [reply compiler/iface/IfaceType.hs]: Without this code, many error messages that used to say `No instance for C ((->) a` say `No instance for C (FUN 'Omega a)`.

File: testsuite/tests/th/T14888.stderr (hunk 7 above)
Comment (monoidal) [reply testsuite/tests/th/T14888.stderr]: AFAICT this change is a casualty of `->` being a type synonym. The presence of kind signature is not relevant to Trac #14888, so I'm just accepting the new output.
